### PR TITLE
 vmm: Streamline reboot code path

### PR DESCRIPTION
**Pull request by sboeuf**
Separate the destruction and cleanup of original VM and the creation of
the new one. In particular have a clear hand off point for resources
(e.g. reset EventFd) used by the new VM from the original. In the
situation where vm.shutdown() generates an error this also avoids the
Vmm reference to the Vm (self.vm) from being maintained.

### CODE REVIEW

File: vmm/src/lib.rs [hunk 1]
```diff
@@ -582,15 +582,16 @@ impl Vmm {
...
             let config = vm.get_config();
             let serial_pty = vm.serial_pty();
             let console_pty = vm.console_pty();
             let console_resize_pipe = vm
                 .console_resize_pipe()
                 .as_ref()
                 .map(|pipe| pipe.try_clone().unwrap());
-            self.vm_shutdown()?;
+            vm.shutdown()?;
+            std::mem::drop(vm);
```
Comment (rbradford) [inline vmm/src/lib.rs:594]: I think calling drop explicitly is a bad idea. The `.take()` will be sufficient to lead to the drop when the block is closed. Perhaps if you're worried about side effects you could close this block immediately after the `vm.shutdown()`.

**Review by rbradford — CHANGES_REQUESTED** [review]

File: vmm/src/lib.rs (hunk 1 above)
Comment (sboeuf) [reply vmm/src/lib.rs:594]: I'm okay with the alternative you're proposing :)

**Comment by sboeuf** [comment]
> I'm not really convinced this makes anything better.

No worries, I don't mind closing the PR.

>You commit message is also wrong the `Vm` is dropped when the `self.vm` parameter is assigned a different value.

No I don't think so, it should be dropped on [line 598](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3858/files#diff-ce16f58dc72b2ebcc4c0720e52982e40e8a7dcdf2669be340f7114dc46b8adb1R598) as you can see this is the end of the if/else statement, from where we do `self.vm.take()`. That means the returned `vm` is only needed until [line 594](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3858/files#diff-ce16f58dc72b2ebcc4c0720e52982e40e8a7dcdf2669be340f7114dc46b8adb1R594).

**Comment by rbradford** [comment]
> > You commit message is also wrong the `Vm` is dropped when the `self.vm` parameter is assigned a different value.
> 
> No I don't think so, it should be dropped on [line 598](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3858/files#diff-ce16f58dc72b2ebcc4c0720e52982e40e8a7dcdf2669be340f7114dc46b8adb1R598) as you can see this is the end of the if/else statement, from where we do `self.vm.take()`. That means the returned `vm` is only needed until [line 594](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3858/files#diff-ce16f58dc72b2ebcc4c0720e52982e40e8a7dcdf2669be340f7114dc46b8adb1R594).

I was referring to your statement in the commit message about how the old code lead to it being dropped.

**Comment by sboeuf** [comment]
> > > You commit message is also wrong the `Vm` is dropped when the `self.vm` parameter is assigned a different value.
> > 
> > 
> > No I don't think so, it should be dropped on [line 598](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3858/files#diff-ce16f58dc72b2ebcc4c0720e52982e40e8a7dcdf2669be340f7114dc46b8adb1R598) as you can see this is the end of the if/else statement, from where we do `self.vm.take()`. That means the returned `vm` is only needed until [line 594](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3858/files#diff-ce16f58dc72b2ebcc4c0720e52982e40e8a7dcdf2669be340f7114dc46b8adb1R594).
> 
> I was referring to your statement in the commit message about how the old code lead to it being dropped.

Gotcha! Well anyway, let's close this PR.

**Comment by rbradford** [comment]
I like the new code but your commit message still needs work. This would be an improvement:

```
vmm: Streamline reboot code path

Separate the destruction and cleanup of original VM and the creation of the new one. In particular have a clear hand off point for resources (e.g. reset EventFd) used by the new VM from the original. In the situation where vm.shutdown() generates an error this also avoids the Vmm reference to the Vm (self.vm) from being maintained.
```

**Comment by sboeuf** [comment]
@rbradford PR's updated!

**Review by rbradford — APPROVED** [review]

**Pull request merged**